chain/network/src/network_protocol/mod.rs

```diff
@@ -31,59 +32,99 @@ use protobuf::Message as _;
 use std::fmt;
 use thiserror::Error;
 
-#[derive(PartialEq, Eq, Clone, Debug)]
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
 pub struct PeerAddr {
     addr: std::net::SocketAddr,
     peer_id: Option<PeerId>,
 }
 
-#[derive(PartialEq, Eq, Clone, Debug)]
-pub struct Validator {
-    peers: Vec<PeerAddr>,
-    account_id: AccountId,
-    epoch_id: EpochId,
-    timestamp: time::Utc,
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub struct AccountData {
+    pub peers: Vec<PeerAddr>,
+    pub account_id: AccountId,
+    pub epoch_id: EpochId,
+    pub timestamp: time::Utc,
 }
 
-impl Validator {
-    pub fn sign(self, signer: &dyn near_crypto::Signer) -> SignedValidator {
+// Limit on the size of the serialized AccountData message.
+// It is important to have such a constraint on the serialized proto,
+// because it may contain many unknown fields (which are dropped during parsing).
+pub const MAX_ACCOUNT_DATA_SIZE_BYTES: usize = 10000; // 10kB
+
+impl AccountData {
+    pub fn sign(self, signer: &dyn near_crypto::Signer) -> anyhow::Result<SignedAccountData> {
         let payload = proto::AccountKeyPayload::from(&self).write_to_bytes().unwrap();
+        if payload.len() > MAX_ACCOUNT_DATA_SIZE_BYTES {
+            anyhow::bail!(
+                "payload size = {}, max is {}",
+                payload.len(),
+                MAX_ACCOUNT_DATA_SIZE_BYTES
+            );
+        }
+        // TODO: here we should validate that payload is not too big - i.e. that it won't exceed
+        // the maximal allowed size.
         let signature = signer.sign(&payload);
-        SignedValidator { validator: self, payload: AccountKeySignedPayload { payload, signature } }
+        Ok(SignedAccountData {
+            account_data: self,
+            payload: AccountKeySignedPayload { payload, signature },
+        })
     }
 }
 
-#[derive(PartialEq, Eq, Clone, Debug)]
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
 pub struct AccountKeySignedPayload {
     payload: Vec<u8>,
     signature: near_crypto::Signature,
 }
 
-#[derive(PartialEq, Eq, Clone, Debug)]
-pub struct SignedValidator {
-    validator: Validator,
+impl AccountKeySignedPayload {
+    pub fn len(&self) -> usize {
+        self.payload.len()
+    }
+    pub fn verify(&self, key: &PublicKey) -> bool {
+        self.signature.verify(&self.payload, key)
+    }
+}
+
+// TODO(gprusak): if we expect this to be large, perhaps we should
+// pass around an Arc, rather than pass it by value.
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub struct SignedAccountData {
+    account_data: AccountData,
     // serialized and signed validator.
```

Comment:
```suggestion
    // serialized and signed by validator.
```

probably also worth mentioning why we need this here (which I assume because we are using protobuf here rather than borsh? I am actually confused here: I was under the impression that we'll continue using `borsh` for stuff that needs signing, did we explicitly decided that that's a good idea to switch from protobufs to borsh even for signable stuff? cc @mm-near)

Reply:
No we didn't, but when it comes to network messages we can always revert that decision.
I wanted to try this out to see how problematic (wrt code maintenance) would it be to have signed protobufs.
I mentioned it in the design doc that I was planning to do that.